src/estimagic/optimization/tranquilo/sample_points.py

```diff
@@ -110,47 +117,207 @@ def _naive_sampler(
+    )
+
+    points = _project_onto_unit_hull(res.params, ord=ord)
+    points = _map_into_feasible_trustregion(points, bounds=effective_bounds)
     return points
 
 
+# ======================================================================================
+# Helper functions
+# ======================================================================================
+
+
+def _pairwise_distance_on_hull(x, existing_xs, ord):  # noqa: A002
+    """Pairwise distance of new and existing points.
+
+    Instead of optimizing the distance of points in the feasible trustregion, this
+    criterion function leads to the maximization of the minimum distance of the points
+    in the unit space. These can then be mapped into the feasible trustregion.
+
+    Args:
+        x (np.ndarray): 2d array of internal points. Each value is in [-1, 1].
+        existing_xs (np.ndarray or None): 2d numpy array in which each row is an
+            x vector at which the criterion function has already been evaluated, that
+            satisfies -1 <= existing_xs <= 1.
+        ord (int): Type of norm to use when scaling the sampled points. For 2 we project
+            onto the hull of a sphere, for np.inf onto the hull of a cube.
+
+    Returns:
+        float: The criterion value.
+
+    """
+    x = _project_onto_unit_hull(x, ord=ord)
+
+    if existing_xs is not None:
+        sample = np.row_stack([x, existing_xs])
+        n_existing_pairs = len(existing_xs) * (len(existing_xs) - 1) // 2
+        slc = slice(0, -n_existing_pairs) if n_existing_pairs else slice(None)
+    else:
+        sample = x
+        slc = slice(None)
+
+    dist = pdist(sample) ** 2  # pairwise squared distances
+    dist = dist[slc]  # drop distances between existing points as they should not
+    # influence the optimization
+
+    crit_value = -logsumexp(-dist)  # smooth minimum
+    return crit_value
+
+
+def _map_into_feasible_trustregion(points, bounds):
+    """Map points from the unit space into trustregion defined by bounds.
+
+    Args:
+        points (np.ndarray): 2d array of points to be mapped. Each value is in [-1, 1].
+        bounds (Bounds): A NamedTuple with attributes ``lower`` and ``upper``, where
+            lower and upper define the rectangle that is the feasible trustregion.
+
+    Returns:
+        np.ndarray: Mapped points.
+
+    """
+    points = (bounds.upper - bounds.lower) * (points + 1) / 2 + bounds.lower
+    return points
+
+
+def _project_onto_unit_hull(x, ord):  # noqa: A002
+    """Project points from the unit space onto the hull of a geometric figure.
+
+    Args:
+        x (np.ndarray): 2d array of points to be projects. Each value is in [-1, 1].
+        ord (int): Type of norm to use when scaling the sampled points. For 2 we project
+            onto the hull of a sphere, for np.inf onto the hull of a cube.
+
+    Returns:
+        np.ndarray: The projected points.
+
+    """
+    norm = np.linalg.norm(x, axis=1, ord=ord).reshape(-1, 1)
```

Comment:
Is the non-differentiability of the infinity norm a problem for the optimizer? Have you tried a high finite order, e.g. 100?

Reply:
It did not seem that non-differentiability was a great problem here I tried a high finite order and it worked as well as the infinity norm. Should we replace the infinity-norm with some large number for ord? Or should we implement a general sampler for an arbitrary order?

Reply:
I would choose an arbitrary order. Only question is then when we switch the default random sampling from normal to uniform. Probably normal for order <= 3 and uniform above that.